### PR TITLE
Explain all and Get all API

### PR DESCRIPTION
**Pull request by bowenlan-amzn**
*Issue #, if available:*
#292 

*Description of changes:*
Explain and Get Policy API now support **get all** functionality, to use it, just don't give any specific query object, directly call GET `_opendistro/_ism/explain` or `_opendistro/_ism/policies`, then you can retrieve all managed indices or all policies.

 Such get all API can have several query parameters, for example:
`GET _opendistro/_ism/policies?from=0&size=20&sortOrder=desc&sortField=policy.last_updated_time&queryString=p*`

Transport action name of get all policies API is `cluster:admin/opendistro/ism/policy/search`
to be consistent with [ODFE permission names](https://opendistro.github.io/for-elasticsearch-docs/docs/security/access-control/permissions/)

### Testing
The IT for Get all policies and Explain All are in `IndexStateManagementRestApiIT.kt` and `RestExplainActionIT.kt`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

---
### Example req and res

`GET _opendistro/_ism/policies`
```
{
  "policies" : [
    {
      "_id" : "test",
      "_seq_no" : 0,
      "_primary_term" : 1,
      "policy" : {
        "policy_id" : "test",
        "description" : "A simple default policy that changes the replica count between hot and cold states.",
        "last_updated_time" : 1611712067239,
        "schema_version" : 1,
        "error_notification" : null,
        "default_state" : "hot",
        "states" : [
          {
            "name" : "hot",
            "actions" : [
              {
                "replica_count" : {
                  "number_of_replicas" : 0
                }
              }
            ],
            "transitions" : [ ]
          }
        ]
      }
    }
  ],
  "total_policies" : 1
}
```

`GET _opendistro/_ism/explain`
```
{
  "testindex1" : {
    "index.opendistro.index_state_management.policy_id" : "test",
    "index" : "testindex1",
    "index_uuid" : "WfnICHPuQ9OlYLXupg96Uw",
    "policy_id" : "test",
    "enabled" : true
  },
  "wildindex" : {
    "index.opendistro.index_state_management.policy_id" : "test",
    "index" : "wildindex",
    "index_uuid" : "47zfTb3JSfe9PGP287glMQ",
    "policy_id" : "test",
    "policy_seq_no" : 0,
    "policy_primary_term" : 1,
    "state" : {
      "name" : "hot",
      "start_time" : 1611712523140
    },
    "retry_info" : {
      "failed" : false,
      "consumed_retries" : 0
    },
    "info" : {
      "message" : "Successfully initialized policy: test"
    },
    "enabled" : true
  },
  "total_managed_indices" : 2
}
```

`GET _opendistro/_ism/explain/testindex1,wildindex`
```
{
  "testindex1" : {
    "index.opendistro.index_state_management.policy_id" : "test",
    "index" : "testindex1",
    "index_uuid" : "WfnICHPuQ9OlYLXupg96Uw",
    "policy_id" : "test",
    "policy_seq_no" : 0,
    "policy_primary_term" : 1,
    "policy_completed" : true
  },
  "wildindex" : {
    "index.opendistro.index_state_management.policy_id" : "test",
    "index" : "wildindex",
    "index_uuid" : "47zfTb3JSfe9PGP287glMQ",
    "policy_id" : "test",
    "policy_seq_no" : 0,
    "policy_primary_term" : 1,
    "policy_completed" : true
  }
}
```

`GET _opendistro/_ism/explain/test*`
```
{
  "testindex1" : {
    "index.opendistro.index_state_management.policy_id" : "test",
    "index" : "testindex1",
    "index_uuid" : "WfnICHPuQ9OlYLXupg96Uw",
    "policy_id" : "test",
    "policy_seq_no" : 0,
    "policy_primary_term" : 1,
    "policy_completed" : true
  }
}

```

`GET _opendistro/_ism/explain?queryString=test*`
total managed indices is determined by search response, according to [previous logic](https://github.com/opendistro-for-elasticsearch/index-management-kibana-plugin/pull/149/files?file-filters%5B%5D=.ts#diff-7a9cad015a905beb60a4d8631efbdabd5178da1abddc06e73a021222d2ce6550L86-L106)
```
{
  "testindex1" : {
    "index.opendistro.index_state_management.policy_id" : "test",
    "index" : "testindex1",
    "index_uuid" : "WfnICHPuQ9OlYLXupg96Uw",
    "policy_id" : "test",
    "policy_seq_no" : 0,
    "policy_primary_term" : 1,
    "state" : {
      "name" : "hot",
      "start_time" : 1611712582258
    },
    "retry_info" : {
      "failed" : false,
      "consumed_retries" : 0
    },
    "info" : {
      "message" : "Successfully initialized policy: test"
    },
    "enabled" : true
  },
  "total_managed_indices" : 1
}
```

### CODE REVIEW

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt [hunk 1]
```diff
@@ -443,7 +443,8 @@ object ManagedIndexRunner : ScheduledJobRunner,
             actionMetaData = null,
             stepMetaData = null,
             policyRetryInfo = PolicyRetryInfoMetaData(failed = true, consumedRetries = 0),
-            info = mapOf("message" to "Fail to load policy: $policyID")
+            info = mapOf("message" to "Fail to load policy: $policyID"),
+            enabled = true
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt:447]: What was this `enabled = true` change for?

Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt:447]: For Kibana side, previously kibana query config index directly for managed-index and get this enabled field

now we can only provide this information through explain API response to Kibana, so add this field.

Comment (dbbaughe) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt:447]: What happens when someone disables the job and this wasn't updated and now metadata says enabled=true and job says enabled=false? You're going to run into a lot of issues trying to keep this in sync with source.

Comment (dbbaughe) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt:447]: As discussed offline, let's not introduce another enabled flag here as keeping it in sync with actual enabled flag is going to introduce issues. Instead just query the config index for the job and get the actual enabled flag value from the source when a user calls the explain API.

File: src/test/resources/mappings/cached-opendistro-ism-history.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "_meta" : {
-    "schema_version": 2
```
Comment (dbbaughe) [inline src/test/resources/mappings/cached-opendistro-ism-history.json:3]: Let's undo changes in this file since we are not adding it to data model

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ManagedIndexMetaData.kt [hunk 3]
```diff
@@ -47,9 +47,11 @@ data class ManagedIndexMetaData(
     val actionMetaData: ActionMetaData?,
     val stepMetaData: StepMetaData?,
     val policyRetryInfo: PolicyRetryInfoMetaData?,
-    val info: Map<String, Any>?
+    val info: Map<String, Any>?,
+    val enabled: Boolean?
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ManagedIndexMetaData.kt:51]: Remove enabled from data model

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/coordinator/ManagedIndexCoordinatorIT.kt [hunk 4]
```diff
@@ -125,7 +125,7 @@ class ManagedIndexCoordinatorIT : IndexStateManagementRestTestCase() {
             assertPredicatesOnMetaData(
                 listOf(index to listOf(ManagedIndexSettings.POLICY_ID.key to fun(policyID: Any?): Boolean = policyID == null)),
                 getExplainMap(index),
-                true
```
Comment (dbbaughe) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/coordinator/ManagedIndexCoordinatorIT.kt:128]: Just curious why the change was needed

Comment (bowenlan-amzn) [reply src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/coordinator/ManagedIndexCoordinatorIT.kt:128]: emmm, not sure anymore, I change it back to true, and it can still pass

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesAction.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.transport.action.getpolicy
+
+import org.elasticsearch.action.ActionType
+
+class GetPoliciesAction private constructor() : ActionType<GetPoliciesResponse>(NAME, ::GetPoliciesResponse) {
+    companion object {
+        val INSTANCE = GetPoliciesAction()
+        val NAME = "cluster:admin/opendistro/ism/policy/read_all"
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesAction.kt:23]: Is this same naming scheme (`read_all`) other plugins using for a GET resources api?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ManagedIndexMetaData.kt [hunk 6]
```diff
@@ -157,6 +159,7 @@ data class ManagedIndexMetaData(
         const val ROLLED_OVER = "rolled_over"
         const val TRANSITION_TO = "transition_to"
         const val INFO = "info"
+        const val ENABLED = "enabled"
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ManagedIndexMetaData.kt:162]: Still need this?

Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ManagedIndexMetaData.kt:162]: ctrl-x malfunction

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesAction.kt (hunk 5 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesAction.kt]: will use `cluster:admin/opendistro/ism/policy/search` to be consistent with others

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/explain/ExplainRequest.kt [hunk 7]
```diff
@@ -28,36 +28,37 @@ class ExplainRequest : ActionRequest {
...
+        searchParams = SearchParams(sin)
     )
 
     override fun validate(): ActionRequestValidationException? {
-        var validationException: ActionRequestValidationException? = null
-        if (indices.isEmpty()) {
-            validationException = ValidateActions.addValidationError("Missing indices", validationException)
-        }
-        return validationException
+        return null
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/explain/ExplainRequest.kt:54]: I am reusing this request for explain single and explain all, for explain all, this can be empty.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesResponse.kt [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                    this.startObject()
+                        .field(_ID, policy.id)
+                        .field(_SEQ_NO, policy.seqNo)
+                        .field(_PRIMARY_TERM, policy.primaryTerm)
+                        .field(Policy.POLICY_TYPE, policy, XCONTENT_WITHOUT_TYPE)
+                        .endObject()
+                }
+            }
+            .endArray()
+            .field("totalPolicies", totalPolicies)
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesResponse.kt:69]: nit: total_policies

**Comment by dbbaughe** [comment]
Can you add example request/responses to the PR description? Want to see what they look like; try adding different examples, e.g. of no indices (all), multiple indices, wildcards etc.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesResponse.kt (hunk 8 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesResponse.kt:69]: changed

**Review by dbbaughe — APPROVED** [review]

**Review by lezzago — APPROVED** [review]
